presto-phoenix/src/main/java/io/prestosql/plugin/phoenix/PhoenixClient.java

```diff
@@ -0,0 +1,453 @@
+                        });
+        }
+        return super.toPrestoType(session, typeHandle);
+    }
+
+    @Override
+    public WriteMapping toWriteMapping(ConnectorSession session, Type type)
+    {
+        if (DOUBLE.equals(type)) {
+            return WriteMapping.doubleMapping("double", doubleWriteFunction());
+        }
+        if (REAL.equals(type)) {
+            return WriteMapping.longMapping("float", realWriteFunction());
+        }
+        if (TIME.equals(type)) {
+            return WriteMapping.longMapping("time", timeWriteFunction());
+        }
+        // Phoenix doesn't support _WITH_TIME_ZONE
+        if (TIME_WITH_TIME_ZONE.equals(type) || TIMESTAMP_WITH_TIME_ZONE.equals(type)) {
+            throw new PrestoException(NOT_SUPPORTED, "Unsupported column type: " + type.getDisplayName());
+        }
+        if (type instanceof ArrayType) {
+            Type elementType = ((ArrayType) type).getElementType();
+            String elementDataType = toWriteMapping(session, elementType).getDataType().toUpperCase();
+            String elementWriteName = getArrayElementPhoenixTypeName(session, this, elementType);
+            return WriteMapping.blockMapping(elementDataType + " ARRAY", arrayWriteFunction(session, elementType, elementWriteName));
+        }
+        return super.toWriteMapping(session, type);
+    }
+
+    private static ColumnMapping arrayColumnMapping(ConnectorSession session, ArrayType arrayType, String elementJdbcTypeName)
+    {
+        return ColumnMapping.blockMapping(
+                arrayType,
+                arrayReadFunction(session, arrayType.getElementType()),
+                arrayWriteFunction(session, arrayType.getElementType(), elementJdbcTypeName));
+    }
+
+    private static BlockReadFunction arrayReadFunction(ConnectorSession session, Type elementType)
+    {
+        return (resultSet, columnIndex) -> {
+            Object[] objectArray = toBoxedArray(resultSet.getArray(columnIndex).getArray());
+            return jdbcObjectArrayToBlock(session, elementType, objectArray);
+        };
+    }
+
+    private static BlockWriteFunction arrayWriteFunction(ConnectorSession session, Type elementType, String elementJdbcTypeName)
+    {
+        return (statement, index, block) -> {
+            Array jdbcArray = statement.getConnection().createArrayOf(elementJdbcTypeName, getJdbcObjectArray(session, elementType, block));
+            statement.setArray(index, jdbcArray);
+        };
+    }
+
+    private JdbcTypeHandle getArrayElementTypeHandle(ConnectorSession session, JdbcTypeHandle arrayTypeHandle)
+    {
+        String arrayTypeName = arrayTypeHandle.getJdbcTypeName()
+                .orElseThrow(() -> new PrestoException(JDBC_ERROR, "Type name is missing: " + arrayTypeHandle));
+        checkArgument(arrayTypeName.endsWith(" ARRAY"), "array type must end with ' ARRAY'");
+        arrayTypeName = arrayTypeName.substring(0, arrayTypeName.lastIndexOf(" ARRAY"));
```

Comment:
We don't need to search since we already know it's at the end
```java
arrayTypeName = arrayTypeName.substring(0, arrayTypeName.length() - " ARRAY".length());